### ISSUE TITLE
Fixed number of subgraphs per datapoint per minibatch

### ISSUE THREAD

**Issue by srn284** [rusty1s/pytorch_geometric#2944]
amazing work on PyG! I had a question:

Say I have 128 data points in a minibatch with some input variables x1, x2, x3. And each datapoint has 200 subgraphs as input in addition to the x1_3. 

Question 1:
How can I design a dataset /data loader to pump out: 128 elements for x1_3 but 25600 (i.e. 200*128) subgraphs for the minibatch?

Question 2:
Some subgraphs are "padding" (ie i have some NN model downstream which has fixed size constraints and thus I need to do some padding as well). How can I design padding subgraphs?

Thank you very much for your time.

**Comment by rusty1s** [comment]
You could potentially group all subgraphs into a single `edge_index` matrix, and hold an additional vector representing the information to which subgraph an edge belongs, e.g.:
```python
edge_indices = torch.cat(edge_indices, dim=-1)
subgraph_type = torch.cat([torch.arange(edge_index.size(1)) for edge_index in edge_indices])
```
which should support mini-batching out-of-the-box. Does that work for you?

For padding, you may want to pad each `data` object in advance, i.e., by filling `x` with zeros. You also want to hold an additional mask of shape `[num_nodes]`, that describes whether a mask is a valid or dummy node.

Hope this helps!

**Comment by srn284** [comment]
Ah thanks very much Matthias! This does help indeed. 

One more question: I end up actually using sequential data for a transformer model in addition to the subgraphs as i discuss above. e.g. 200 tokens per datapoint and 200 subgraphs for datapoint. How do I go about storing this in a dataset? I tried inmemorydataset, and it just crashes with memory error at the self.collate phase of the process() method. A lot of data I imagine.

I tried the solution presented for larger data. I save it one by one in a folder and develop a get method to load it. But after all points have been saved, initialising the dataset takes ages. Not sure why. 

Any potential solutions?

```

import torch
from torch_geometric.data import InMemoryDataset
from tqdm import tqdm
from torch_geometric.data import Data, Dataset

class RPLoader_EHR_KG_SeqGR_big(Dataset):
    def __init__(self, root, savefolder = None, transform=None, pre_transform=None, token2idx=None, max_len=400, max_age=110, year=False, age_symbol=None, min_visit=5,
                 yearOn=False, yearV=None, KG=None, pureICD=None,
                 maxModality=10, maskToken=None, breakpt=200 ):
        self.vocab = token2idx
        self.max_len = max_len
        
        self.age2idx, _ = age_vocab(max_age, year, symbol=age_symbol)

        self.yearOn = yearOn
        self.yearV = yearV
        self.KG = KG
        self.pureICD = pureICD['token2idx']
        self.maxModality = maxModality
        self.maskTokenKG = maskToken
        self.breakpt = breakpt
        self.EHRmaxlen = self.breakpt
        self.KGmaxlen = self.max_len - self.breakpt
        self.maxTok = int(self.KGmaxlen )
        self.savefolder = savefolder
        print('inits done...')
        
        
        super(RPLoader_EHR_KG_SeqGR_big, self).__init__(root, transform, pre_transform)
        

    @property
    def raw_file_names(self):
        return [   self.root + '/HF_BEHRTgen_randomCut_40_2005_min36.parquet']
    @property
    def processed_file_names(self):
        
        return [self.root+"/"+self.savefolder+'data_'+str(index)+".pt" for index in range(636511)]
#     [self.root + '/HF_Data.dataset'] 

    def download(self):
        pass
    def getGraph(self, code, age, position, segment, year):
        #   do a bunch of stuff to extract nodes and extract edge index (written below as KGgraph)

        nodes = [np.array(list(self.KG[self.pureICD[DiagEl]][0])) for DiagEl in codeOut]
        KGgraph = [list(self.KG[self.pureICD[DiagEl]][1] )for DiagEl in codeOut]
        item_encoder = LabelEncoder()

        cumsum = 0

        KGgraphNorm_from=[]
        KGgraphNorm_to=[]
        seqOut = []
        for subi, subgraph in enumerate(nodes):
            item_encoder = LabelEncoder()

            nodeSub =item_encoder.fit_transform(subgraph) 
#             print('nodesnorm', nodeSub)
            lookupNode = dict(zip(subgraph, nodeSub))

            subG = KGgraph[subi]
            edge_index_from = np.array([lookupNode[subGel[0]] for subGel in subG])
            edge_index_to = np.array([lookupNode[subGel[1]] for subGel in subG])
#             edge_index = np.array([edge_index_from, edge_index_to])
#             print(edge_index)
            KGgraphNorm_from.append(edge_index_from + cumsum)
            KGgraphNorm_to.append(edge_index_to + cumsum)
            seqOut.append([subi]*len(lookupNode))
            cumsum = cumsum + len(lookupNode)
#         print('edge norms... ', KGgraphNorm)
        
        KGgraphNorm_from = np.hstack(np.array(KGgraphNorm_from))
        KGgraphNorm_to = np.hstack(np.array(KGgraphNorm_to))


        KGgraphNorm = torch.LongTensor([KGgraphNorm_from, KGgraphNorm_to])  

        nodesNorm = torch.LongTensor(np.hstack(np.array(nodes))).unsqueeze(-1)

        seqOut = np.hstack(np.array(seqOut))
        
        return nodesNorm, KGgraphNorm, seqOut

    def process(self):
        
        data_list = []
        print(self.raw_file_names[0])
        # process by session_id
        df = pd.read_parquet(self.raw_file_names[0])
        
        df['label'] = df.label.apply(lambda x : int(x))
#         groups = df.groupby('patid')

        for index, roww in tqdm(df.iterrows()):
            

                # do a bunch of stuff to get the age, code, position, segment, year all important for sequential modelling. 
                # i derive the graph per datapoint from this information.in self.getGraph() below....

                nodes, KG, seq = self.getGraph(pureCode, pureAge, position, segment, year)

                data = Data(age= age, code= code, position=position, region = torch.LongTensor([region]), gender = torch.LongTensor([gender]), label = label, year=year, x= nodes, edge_index = KG)
                data.seq= seq
                torch.save(data,self.root+"/"+self.savefolder+'data_'+str(index)+".pt")
    def len(self):
        return len(self.processed_file_names)

    def get(self, idx):
        data = torch.load(self.root+"/"+self.savefolder+'data_'+str(idx)+".pt")
        return data
```

**Comment by rusty1s** [comment]
Which part takes ages? Initializing or iterating? Initializing should be instantaneously, as there is nothing to load from disk at this stage.

**Comment by srn284** [comment]
In the initializing part; that's what is so weird! 

I did a debug trace and it takes years on the files_exist method called in the _process method in the Dataset class:


```
        if files_exist(self.processed_paths):  # pragma: no cover
            return

```


Not sure if I am doing something wrong (in the code posted above in my last message) or something naturally slow about reading in these graphs? Thanks again for your help Matthias.

**Comment by rusty1s** [comment]
Yes, it will check for the existence of every of those 636511 files, which is probably not needed. Try to just check for a bunch of them:
```python
    @property
    def processed_file_names(self):
        return [self.root+"/"+self.savefolder+'data_'+str(index)+".pt" for index in range(10)]
```

**Comment by srn284** [comment]
Fantastic, I solved it actually by throwing out the process method over the weekend, but this is easier. Thanks much Matthias for your speedy responses and resolution :)
